File: Advanced/PageAndRDL/CustomPdfExport/C#/TestApplication/FileStreamProvider.cs
using System;
using System.IO;
using GrapeCity.ActiveReports.Extensibility.Rendering.IO;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering
{
	internal sealed class FileStreamProvider : StreamProvider
	{
		private readonly string _outputPath;
		private StreamInfo _primaryStream;

		public FileStreamProvider(string outputPath)
		{
			_outputPath = outputPath;
		}

		#region Overrides of StreamProvider

		public override StreamInfo CreatePrimaryStream(string mimeType, string suggestedFileExtension)
		{
			return _primaryStream = new FileStreamInfo(new Uri(_outputPath, UriKind.RelativeOrAbsolute), mimeType);
		}

		public override StreamInfo CreateSecondaryStream(string relativeName, string mimeType, string suggestedFileExtension)
		{
			throw new NotSupportedException();
		}

		public override StreamInfo GetPrimaryStream()
		{
			return _primaryStream;
		}

		public override StreamInfo[] GetSecondaryStreams()
		{
			throw new NotSupportedException();
		}

		#endregion
		
		private sealed class FileStreamInfo : StreamInfo
		{
			public FileStreamInfo(Uri uri, string mimeType) : base(uri, mimeType)
			{
			}
			
			public override Stream OpenStream()
			{
				return new FileStream(Uri.OriginalString, FileMode.Create, FileAccess.ReadWrite, FileShare.None);
			}
		}
	}
}

File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/TestApplication/FileStreamProvider.vb
Imports System.IO
Imports System
Imports GrapeCity.ActiveReports.Extensibility.Rendering.IO

Friend NotInheritable Class FileStreamProvider
	Inherits StreamProvider
	Private ReadOnly _outputPath As String
	Private _primaryStream As StreamInfo

	Public Sub New(outputPath As String)
		_outputPath = outputPath
	End Sub

#Region "Overrides of StreamProvider"

	Public Overrides Function CreatePrimaryStream(mimeType As String, suggestedFileExtension As String) As StreamInfo
		_primaryStream = New FileStreamInfo(New Uri(_outputPath, UriKind.RelativeOrAbsolute), mimeType)
		Return _primaryStream
	End Function

	Public Overrides Function CreateSecondaryStream(relativeName As String, mimeType As String, suggestedFileExtension As String) As StreamInfo
		Throw New NotSupportedException()
	End Function

	Public Overrides Function GetPrimaryStream() As StreamInfo
		Return _primaryStream
	End Function

	Public Overrides Function GetSecondaryStreams() As StreamInfo()
		Throw New NotSupportedException()
	End Function

#End Region

	Private NotInheritable Class FileStreamInfo
		Inherits StreamInfo
		Public Sub New(uri As Uri, mimeType As String)
			MyBase.New(uri, mimeType)
		End Sub

		Public Overrides Function OpenStream() As Stream
			Return New FileStream(Uri.OriginalString, FileMode.Create, FileAccess.ReadWrite, FileShare.None)
		End Function
	End Class
End Class
